code/game/objects/items/robot/robot_upgrades.dm

```diff
@@ -343,92 +402,81 @@
-
-/obj/item/borg/upgrade/selfrepair/process()
-	if(world.time < next_repair)
-		return
 
-	if(cyborg && (cyborg.stat != DEAD) && on)
-		if(!cyborg.cell)
-			to_chat(cyborg, "<span class='warning'>Self-repair module deactivated. Please, insert the power cell.</span>")
-			deactivate_sr()
-			return
-
-		if(cyborg.cell.charge < powercost * 2)
-			to_chat(cyborg, "<span class='warning'>Self-repair module deactivated. Please recharge.</span>")
-			deactivate_sr()
-			return
-
-		if(cyborg.health < cyborg.maxHealth)
-			if(cyborg.health < 0)
-				repair_amount = -2.5
-				powercost = 30
-			else
-				repair_amount = -1
-				powercost = 10
-			cyborg.adjustBruteLoss(repair_amount)
+/obj/item/borg/upgrade/selfrepair/proc/cooldown_start(ticksleft = 0)
+	icon_state = "selfrepair_off"
+	update_icon()
+	playsound(cyborg.loc, deactivation_sound, 60)
+	COOLDOWN_START(src, recharging_repmod, ((20 SECONDS) + ((100 SECONDS) - ((10 SECONDS) * ticksleft)))) //20 seconds + 10 seconds for every repair tick used up. 2 minutes max
+	begin_timer_animation()
+
+/obj/item/borg/upgrade/selfrepair/proc/startrepair()
+	while(repair_ticks)
+		sleep(10)
+		if(!cyborg)
+			return FALSE
+		if(!cyborg.cell || cyborg.cell.charge <= powercost * 10)
+			repair_ticks = 0
+			to_chat(cyborg, "<span class='notice'>Power level critically low! Your self-repair module has deactivated early.</span>")
+		else if(cyborg.getFireLoss())
+			playsound(cyborg.loc, pick(running_sounds), 30)
 			cyborg.adjustFireLoss(repair_amount)
-			cyborg.updatehealth()
 			cyborg.cell.use(powercost)
+			repair_ticks--
+		else if(cyborg.getBruteLoss())
+			playsound(cyborg.loc, pick(running_sounds), 30)
+			cyborg.adjustBruteLoss(repair_amount)
+			cyborg.cell.use(powercost)
+			repair_ticks--
 		else
-			cyborg.cell.use(5)
-		next_repair = world.time + repair_cooldown * 10 // Multiply by 10 since world.time is in deciseconds
-
-		if(!TIMER_COOLDOWN_CHECK(src, COOLDOWN_BORG_SELF_REPAIR))
-			TIMER_COOLDOWN_START(src, COOLDOWN_BORG_SELF_REPAIR, 200 SECONDS)
-			var/msgmode = "standby"
-			if(cyborg.health < 0)
-				msgmode = "critical"
-			else if(cyborg.health < cyborg.maxHealth)
-				msgmode = "normal"
-			to_chat(cyborg, "<span class='notice'>Self-repair is active in <span class='boldnotice'>[msgmode]</span> mode.</span>")
-	else
+			to_chat(cyborg, "<span class='notice'>You are fully repaired, your module has deactivated early.</span>")
+			cooldown_start(repair_ticks)
+			repair_ticks = 0
+	cooldown_start(0)
+
+/obj/item/borg/upgrade/selfrepair/process(delta_time)
+	if(!cyborg) //Sanity check to reset the module in case it is somehow removed while running.
+		update_icon()
+		COOLDOWN_RESET(src, recharging_repmod)
+		repair_ticks = 10
 		deactivate_sr()
+	if(!COOLDOWN_FINISHED(src, recharging_repmod))
+		update_timer_animation()
+	if(COOLDOWN_FINISHED(src, recharging_repmod) && icon_state == "selfrepair_off")
+		end_timer_animation()
+		action.UpdateButtonIcon()
+		repair_ticks = 10
+		deactivate_sr()
```

Comment:
Account for delta time. Healing rates need to be multiplied by delta_time, so that if the subsystem is slowed down healing will remain the same per second.